python/tk_desktop/desktop_window.py

```diff
@@ -600,6 +605,64 @@ def _register_apps_tab(self):
     def contextMenuEvent(self, event):
         self.user_menu.exec_(event.globalPos())
 
+    def closeEvent(self, event):
+        # When closing the app programatically via self.close(),
+        # the event is not spontaenous and we know the state of self._is_quitting
+        # is correct.
+
+        # When CMD-Q is hit or the close button is pressed however, the dialog
+        # first receives a closeEvent and THEN receives the aboutToQuit signal,
+        # which will call the close event. When this happens, the event is marked
+        # as spontaneous, but we can't tell what actually happened so we postpone
+        # the execution of it by making sure it gets called again after the event
+        # queue is emptied.
+        # When the timer has expired, we will know if the spontaneous event had
+        # been a simple close of the dialog OR if the app was closed.
+        # If is_quitting isn't set by that point, it's because the user pressed
+        # the close button. If it is, then we know the app is closing and we know
+        # how to deal with that.
+        if event.spontaneous() is False or self._is_quitting:
+            self._handle_close_event()
+        else:
+            QtCore.QTimer.singleShot(0, self._handle_close_event)
+
+    def _handle_close_event(self):
+        if self._is_quitting:
```

Comment:
So this method, if we assume for a sec that _is_quitting is false, will just handle the minimize of the app and not the shutting down?

Reply:
Maybe it should have a doc string?